tests/core/permissions/permissions.cpp

```diff
@@ -0,0 +1,366 @@
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * permissions.cpp -- unit test for pmemfile_chmod, chown & co
+ */
+#include "pmemfile_test.hpp"
+#include <cstring>
+
+class permissions : public pmemfile_test {
+public:
+	permissions() : pmemfile_test()
+	{
+	}
+};
+
+TEST_F(permissions, chmod)
+{
+	PMEMfile *f;
+	struct stat statbuf;
+
+	ASSERT_TRUE(test_pmemfile_create(pfp, "/aaa", PMEMFILE_O_EXCL,
+					 PMEMFILE_S_IRUSR | PMEMFILE_S_IWUSR |
+						 PMEMFILE_S_IRGRP |
+						 PMEMFILE_S_IROTH));
+	ASSERT_EQ(pmemfile_stat(pfp, "/aaa", &statbuf), 0);
+	EXPECT_EQ(statbuf.st_mode & PMEMFILE_ALLPERMS,
+		  (mode_t)(PMEMFILE_S_IRUSR | PMEMFILE_S_IWUSR |
+			   PMEMFILE_S_IRGRP | PMEMFILE_S_IROTH));
+
+	ASSERT_EQ(pmemfile_chmod(pfp, "/a_not_exists",
+				 PMEMFILE_S_IRUSR | PMEMFILE_S_IWUSR),
+		  -1);
+	EXPECT_EQ(errno, ENOENT);
+
+	/*
+	 * Adding group or other permissions should not change anything
+	 * WRT permission checks.
+	 */
+	for (const auto &m :
+	     std::vector<mode_t>{0, PMEMFILE_S_IRGRP | PMEMFILE_S_IWGRP,
+				 PMEMFILE_S_IROTH | PMEMFILE_S_IWOTH}) {
+		SCOPED_TRACE(m);
+
+		/* chmod u+rw */
+		ASSERT_EQ(pmemfile_chmod(pfp, "/aaa", PMEMFILE_S_IRUSR |
+						 PMEMFILE_S_IWUSR | m),
+			  0)
+			<< strerror(errno);
+		ASSERT_EQ(pmemfile_stat(pfp, "/aaa", &statbuf), 0);
+		EXPECT_EQ(statbuf.st_mode & PMEMFILE_ALLPERMS,
+			  (mode_t)(PMEMFILE_S_IRUSR | PMEMFILE_S_IWUSR | m));
+
+		/* open rw */
+		f = pmemfile_open(pfp, "/aaa", PMEMFILE_O_RDWR);
+		ASSERT_NE(f, nullptr) << strerror(errno);
+		pmemfile_close(pfp, f);
+
+		/* open r */
+		f = pmemfile_open(pfp, "/aaa", PMEMFILE_O_RDONLY);
+		ASSERT_NE(f, nullptr) << strerror(errno);
+		pmemfile_close(pfp, f);
+
+		/* open w */
+		f = pmemfile_open(pfp, "/aaa", PMEMFILE_O_WRONLY);
+		ASSERT_NE(f, nullptr) << strerror(errno);
+		pmemfile_close(pfp, f);
+	}
+
+	for (const auto &m :
```

Comment:
Or rather:

for (mode_t m : {0, PMEMFILE_S_IRGRP | PMEMFILE_S_IWGRP,
				 PMEMFILE_S_IROTH | PMEMFILE_S_IWOTH})